mlrun/k8s_utils.py

```diff
@@ -228,3 +230,322 @@ def handle_invalid(message):
+    for toleration in to_add:
+        if toleration not in merged:
+            merged.append(toleration)
+    return merged
+
+
+def _prune_node_selector(
+    node_selector: dict[str, str],
+    enriched_node_selector: dict[str, str],
+):
+    """
+    Prunes given node_selector key from function spec if their key and value are matching
+    :param node_selector: node selectors to prune
+    """
+    # both needs to exists to prune required node_selector from the spec node selector
+    if not node_selector or not enriched_node_selector:
+        return
+
+    mlrun.utils.logger.debug("Pruning node selectors", node_selector=node_selector)
+    for key, value in node_selector.items():
+        if value:
+            spec_value = enriched_node_selector.get(key)
+            if spec_value and spec_value == value:
+                enriched_node_selector.pop(key)
+
+    return enriched_node_selector
+
+
+def _prune_affinity_node_selector_requirement(
+    node_selector_requirements: list[kubernetes.client.V1NodeSelectorRequirement],
+    affinity: typing.Optional[kubernetes.client.V1Affinity],
+):
+    """
+    Prunes given node selector requirements from affinity.
+    We are only editing required_during_scheduling_ignored_during_execution because the scheduler can't schedule
+    the pod unless the rule is met.
+    :param node_selector_requirements:
+    :return:
+    """
+    # both needs to exist to prune required affinity from spec affinity
+    if not affinity or not node_selector_requirements:
+        return
+    if affinity.node_affinity:
+        node_affinity: kubernetes.client.V1NodeAffinity = affinity.node_affinity
+
+        new_required_during_scheduling_ignored_during_execution = None
+        if node_affinity.required_during_scheduling_ignored_during_execution:
+            node_selector: kubernetes.client.V1NodeSelector = (
+                node_affinity.required_during_scheduling_ignored_during_execution
+            )
+            new_node_selector_terms = (
+                _prune_node_selector_requirements_from_node_selector_terms(
+                    node_selector_terms=node_selector.node_selector_terms,
+                    requirements_to_prune=node_selector_requirements,
+                )
+            )
+            # check whether there are node selector terms to add to the new list of required terms
+            if len(new_node_selector_terms) > 0:
+                new_required_during_scheduling_ignored_during_execution = (
+                    kubernetes.client.V1NodeSelector(
+                        node_selector_terms=new_node_selector_terms
+                    )
+                )
+        # if both preferred and new required are empty, clean node_affinity
+        if (
+            not node_affinity.preferred_during_scheduling_ignored_during_execution
+            and not new_required_during_scheduling_ignored_during_execution
+        ):
+            setattr(affinity, "node_affinity", None)
```

Comment:
Is there a reason not to do this here?
```suggestion
            affinity.node_affinity = None
```